dictzip.1
.\" dictzip.1 -- 
.\" Created: Sun Jun 22 16:33:39 1997 by faith@acm.org
.\" Revised: Fri Jul 11 14:18:25 1997 by faith@acm.org
.\" Copyright 1997 Rickard E. Faith (faith@acm.org)
.\" 
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\" 
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\" 
.TH DICTZIP 1 "22 Jun 1997" "" "Linux Programmer's Manual"
.SH "NAME"
dictzip, dictunzip \- compress (or expand) files, allowing random access
.SH "SYNOPSIS"
.BI "dictzip [" options "] " name
.br
.BI "dictunzip [" options "] " name
.SH "DESCRIPTION"
.B dictzip
compresses files using the
.BR gzip (1)
algorithm (LZ77) in a manner which is completely compatible with the
.BR gzip
file format.  An extension to the
.B gzip
file format (Extra Field, described in 2.3.1.1 of RFC 1952) allows extra
data to be stored in the header of a compressed file.  Programs like
.B gzip
and
.B zcat
will ignore this extra data.  However,
.BR dictd (8),
the DICT protocol dictionary server will make use of this data to perform
pseudo-random access on the file.  Files in the
.B dictzip
format should end in ".dz" so that they may be distinguished from common
.B gzip
files that do not contain the special header information.
.P
From RFC 1952, the extra field is specified as follows:
.sp
.RS
If the FLG.FEXTRA bit is set, an "extra field" is present in
the header, with total length XLEN bytes.  It consists of a
series of subfields, each of the form:
.sp
.nf
+---+---+---+---+==================================+
|SI1|SI2|  LEN  |... LEN bytes of subfield data ...|
+---+---+---+---+==================================+
.fi
.sp
SI1 and SI2 provide a subfield ID, typically two ASCII letters
with some mnemonic value.  Jean-Loup Gailly
<gzip@prep.ai.mit.edu> is maintaining a registry of subfield
IDs; please send him any subfield ID you wish to use.  Subfield
IDs with SI2 = 0 are reserved for future use.
.P
LEN gives the length of the subfield data, excluding the 4
initial bytes.
.RE
.sp
The
.B dictzip
program uses 'R' for SI1, and 'A' for SI2 (i.e., "Random Access").  After
the LEN field, the data is arranged as follows:
.sp
.nf
+---+---+---+---+---+---+===============================+
|  VER  | CHLEN | CHCNT |  ... CHCNT words of data ...  |
+---+---+---+---+---+---+===============================+
.fi
.sp
As per RFC 1952, all data is stored least-significant byte first.  For VER
1 of the data, all values are 16-bits long (2 bytes), and are unsigned
integers.
.P
XLEN (which is specified earlier in the header) is a two byte integer, so
the extra field can be 0xffff bytes long, 2 bytes of which are used for the
subfield ID (SI1 and SI1), and 2 bytes of which are used for the subfield
length (LEN).  This leaves 0xfffb bytes (0x7ffd 2-byte entries or 0x3ffe
4-byte entries).  Given that the zip output buffer must be 10% + 12 bytes
larger than the input buffer, we can store 58969 bytes per entry, or about
1.8GB if the 2-byte entries are used.  If this becomes a limiting factor,
another format version can be selected and defined for 4-byte entries.
.P
For compression, the file is divided up into "chunks" of data, each chunk
is less than 64kB, and can be compressed into an area that is also less
than 64kB long (taking incompressible data into account -- usually the data
is compressed into a block that is much smaller than the original).  The
CHLEN field specifies the length of a "chunk" of data.  The CHCNT field
specifies how many chunks are preset, and the CHCNT words of data specifies
how long each chunk is after compression (i.e., in the current compressed
file).
.P
To perform random access on the data, the offset and length of the data are
provided to library routines.  These routines determine the chunk in which
the desired data begins, and decompresses that chunk.  Consecutive chunks
are decompressed as necessary.
.SH "TRADEOFFS"
.TP
.B Speed
True random file access is not realized, since any access, even for a
single byte, requires that a 64kB chunk be read and decompressed.  This is
slower than accessing a flat text file, but is much, much faster than
performing serial access on a fully compressed file.
.TP
.B Space
For the textual dictionary databases we are working with, the use of 64kB
chunks and maximal LZ77 compression realizes a file which is only about 4%
larger than the same file compressed all at once.
.SH "OPTIONS"
.TP
.BR \-d " or " \-\-decompress
Decompress.  This is the default if the executable is called
.BR dictunzip .
.TP
.BR \-c " or " \-\-stdout
Write output on standard output; keep original files unchanged.  This is
only available when decompressing (because parts of the header must be
updated after a write when compressing).
.TP
.BR \-f " or " \-\-force
Force compression or decompression even if the output file already exists.
.TP
.BR \-h " or " \-\-help
Display help.
.TP
.BR \-k " or " \-\-keep
Do not delete the original file.
.TP
.BR \-l " or " \-\-list
For each compressed file, list the following fields:

    type: dzip, gzip, or text (includes files in unknown formats)
    crc: CRC checksum
    date and time: from header
    chunks: number of chunks in file
    size: size of each uncompressed chunk
    compr.: compressed size
    uncompr.: uncompressed size
    ratio: compression ratio (0.0% if unknown)
    name: name of uncompressed file

Unlike
.BR gzip ,
the compression method is not detected.
.TP
.BR \-L " or " \-\-license
Display the
.B dictzip
license and quit.
.TP
.BR \-t " or " \-\-test
Check the compressed file integrity.  This option is not implemented.
Instead, it will list the header information.
.TP
.BR \-v " or " \-\-verbose
Verbose. Display extra information during compression.
.TP
.BR \-V " or " \-\-version
Version. Display the version number and compilation options then quit.
.TP
.BI \-s " start\fR or " \-\-start " start"
Specify the offer to start decompression, using decimal numbers.  The
default is at the beginning of the file.
.TP
.BI \-e " size\fR or " \-\-size " size"
Specify the size of the portion of the file to decompress, using decimal
numbers.  The default is the whole file.
.TP
.BI \-S " start\fR or " \-\-Start " start"
Specify the offer to start decompression, using base64 numbers.  The
default is at the beginning of the file.
.TP
.BI \-E " size\fR or " \-\-Size " start"
Specify the size of the portion of the file to decompress, using base64
numbers.  The default is the whole file.
.TP
.BI \-p " prefilter\fR or " \-\-pre " prefilter"
Specify a shell command to execute as a filter before compression or
decompression of a chunk.  The pre- and post-compression filters can be
used to provide additional compression or output formatting.  The filters
may not increase the buffer size significantly.  The pre- and
post-compression filters were designed to provide the most general
interface possible.
.TP
.BI \-P " postfilter\fR or " \-\-post " postfilter"
Specify a shell command to execute as a filter after compression or
decompression.
.SH CREDITS
.B dictzip
was written by Rik Faith (faith@cs.unc.edu) and is distributed under the
terms of the GNU General Public License.  If you need to distribute under
other terms, write to the author.
.P
The main libraries used by this programs (zlib, regex, libmaa) are
distributed under different terms, so you may be able to use the libraries
for applications which are incompatible with the GPL -- please see the
copyright notices and license information that come with the libraries for
more information, and consult with your attorney to resolve these issues.
.SH "SEE ALSO"
.BR dict (1),
.BR dictd (8),
.BR gzip (1),
.BR gunzip (1),
.BR zcat (1)


NEWS
Version 1.9.13, by Aleksey Cheusov, 24 Mar 2004 11:32:53 +0200

  FIXED: if database has no 00-database-alphabet headword,
      "global" alphabet is built incorrectly. As a result
      dictd may work (FreeBSD does) incorrectly with LEV strategy.
      Affected versions: 1.9.12

  ADDED: plugin using libdbi library for implementing DICT database using
    SQL server. It is possible to specify custom search strategies
    by specifying SQL query.

    The following option are currently supported:

      driverdir       - path to DBI drivers
      drivername      - DBI driver name, such as pgsql, mysql, sqlite etc.
      option_host     - host to connect to
      option_port     - port to connect to
      option_dbname   - database name
      option_username - user name for authorizing to SQL server
      option_password - user password for authorizing to SQL server
      query_define    -        SQL query for implementing DEFINE command
      query_<strategy> - SQL query for implementing
           'strategy' (MATCH command).
           Note that 'strategy' is not limited to the "standard" ones.

    See the file example_plugin_dbi.conf for sample of how
      this plugin can be configured.

    SQL queries should return 1-column result by using SELECT command.
      %q sequence inside SQL query is expanded to user's query.
      %% is expanded to single % sign.

    This plugin gives a simplest way to implement editable
      dictionaries with a help
      of WEB interface, but with a read-only access to it
      using DICT protocol.

    NOTE! Unless libdbi library is not found on your system
    or you configure dictd with --disable-plugin option,
    'dictd' is linked with libdbi to resolv DBI driver's symbols.
    This should be fixed.

    NOTE! Connection to SQL database is made
    for every DEFINE/MATCH command.
    This slows down plugin and should be fixed.

Version 1.9.12, by Aleksey Cheusov, 10 Mar 2004 18:45:15 +0200

  Both dictd and Judy-plugin correctly implement
    the 'lev' strategy for utf-8 databases.
    For this purpose an additional headword 00-database-alphabet is
    created by dictfmt.
    So, if you often use 'lev' strategy with utf-8 dictionaries,
    rebuild them using dictfmt-1.9.12

  'nl_langinfo' function is used to detect
      whether dictd/dictfmt uses 8bit, utf-8, or C locale.

  dictd:
    - fixes

       SHOW SERVER command cuts off long database names

       crash on Solaris (and FreeBSD5 ?) due to
         toupper('255') returning 376 if utf-8 locale is set.

       LC_CTYPE and LC_COLLATE are set according to --locale.
         As a result date and time are logged using the C locale

    - MATCH|DEFINE *|! ignore invisible dictionaries.
         Now there is no reason to separate visible and invisible dictionaries
         by database_exit command.
         This simplifies the configuration file.

    - dictd --test|--test-file notifies about invalid database name

  dictl:

    - checks whether recode/iconv/konwert/locale commands are available

    - analyses PAGER variable to set a pager.

    - incorrect charset determination has been fixed

  dictfmt:

     - new option --break-headwords for use with --headword-separator, so
       each headword is written on its own line of the .dict file.

     - correctly wraps utf-8 input

     - new option --columns to set a number of columns for
      wrapping text. See dictfmt.1. By default this value is 72.
      All formats are affected and use the same default.

     - 00-database-XXX headwords are not copied to .dict file.
         The only exception is 00-database-short which should be copied
         to .dict file for older 'dictd'

  maa:
      new functions and #defines:
         'str_pool_init_position', 'str_pool_next_position',
         'str_pool_get_position', 'str_pool_readonly', 'str_pool_iterate',
         'str_pool_iterate_arg', 'STR_ITERATE' and 'STR_ITERATE_END'
         to iterate over strings in the string pool.

      fixed a bug in flags.c/flg_set function

  dictdplugin_judy:
      
      fixed: if a requested headword has more than one definition,
         DEFINE command always returns the first one

Version 1.9.11, by Aleksey Cheusov, 8 Dec 2003 20:55:31 +0200

  NEWS file added

  dictd:
    - dictd is compiled successfully on Solaris, MacOS-X(cc).
      A lot of warning and error messages appeared
      on 64-bit Opteron and Alpha were fixed.

    - fixed: 'suffix' strategy doesn't ignore characters other than
       alphanumeric and spaces. Affected versions: 1.9.1-1.9.10
    - A multi-lined short name of the database is truncated before sending
      to the client

    - patch by Michael Bunk <bunk at imn.htwk-leipzig.de} fixing:
       - a typo giving the incomprehensible failure message.
         Affected versions: 1.7.1-1.9.10

       - a typo resulting in 'dict --server-info' showing
           same compressed/uncompressed database sizes
         Affected versions: 1.9.1-1.9.10

       - broken AUTH command.
         Affected versions: 1.9.4-1.9.10

    - added: keyword "disable_strategy for disabling a specified strategy
             for a particular database.
             Strategies disabled by 'disable_strategy' keyword
             are not passed to plugins.

    - info/name keywords in 'database_virtual' section can handle
      entry name beginning with @ symbol in a similar way
      as 'database' and 'database_plugin' do.

    - dictd --pp <prog> sets a preprocessor for configuration file.

      If you run DICT server having a lot of dictionaries, you can preprocess
      configuration file using m4 (or cpp). This significantly reduces
      the size of configuration file.

      See dictd.8 and example_complex.conf.

  plugins:
    - the default directory for databases

dictzip.c
/* dictzip.c -- 
 * Created: Tue Jul 16 12:45:41 1996 by faith@dict.org
 * Revised: Mon Apr 22 11:40:17 2002 by faith@dict.org
 * Copyright 1996-1998, 2000, 2002 Rickard E. Faith (faith@dict.org)
 * 
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 1, or (at your option) any
 * later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include "dictzip.h"
#include "data.h"

#include <sys/stat.h>

void dict_data_print_header( FILE *str, dictData *header )
{
   char        *date, *year;
   long        ratio, num, den;
   static int  first = 1;

   if (first) {
      fprintf( str,
	       "type   crc        date    time chunks  size     compr."
	       "  uncompr. ratio name\n" );
      first = 0;
   }
   
   switch (header->type) {
   case DICT_TEXT:
      date = ctime( &header->mtime ) + 4; /* no day of week */
      date[12] = date[20] = '\0'; /* no year or newline*/
      year = &date[16];
      fprintf( str, "text %08lx %s %11s ", header->crc, year, date );
      fprintf( str, "            " );
      fprintf( str, "          %9ld ", header->length );
      fprintf( str, "  0.0%% %s",
	       header->origFilename ? header->origFilename : "" );
      putc( '\n', str );
      break;
   case DICT_GZIP:
   case DICT_DZIP:
      fprintf( str, "%s", header->type == DICT_DZIP ? "dzip " : "gzip " );
#if 0
      switch (header->method) {
      case 0:  fprintf( str, "store" ); break;
      case 1:  fprintf( str, "compr" ); break;
      case 2:  fprintf( str, "pack " ); break;
      case 3:  fprintf( str, "lzh  " ); break;
      case 8:  fprintf( str, "defla" ); break;
      default: fprintf( str, "?    " ); break;
      }
#endif
      date = ctime( &header->mtime ) + 4; /* no day of week */
      date[12] = date[20] = '\0'; /* no year or newline*/
      year = &date[16];
      fprintf( str, "%08lx %s %11s ", header->crc, year, date );
      if (header->type == DICT_DZIP) {
	 fprintf( str, "%5d %5d ", header->chunkCount, header->chunkLength );
      } else {
	 fprintf( str, "            " );
      }
      fprintf( str, "%9ld %9ld ", header->compressedLength, header->length );
      /* Algorithm for calculating ratio from gzip-1.2.4,
         util.c:display_ratio Copyright (C) 1992-1993 Jean-loup Gailly.
         May be distributed under the terms of the GNU General Public
         License. */
      num = header->length-(header->compressedLength-header->headerLength);
      den = header->length;
      if (!den)
	 ratio = 0;
      else if (den < 2147483L)
	 ratio = 1000L * num / den;
      else
	 ratio = num / (den/1000L);
      if (ratio < 0) {
	 putc( '-', str );
	 ratio = -ratio;
      } else putc( ' ', str );
      fprintf( str, "%2ld.%1ld%%", ratio / 10L, ratio % 10L);
      fprintf( str, " %s",
	       header->origFilename ? header->origFilename : "" );
      putc( '\n', str );
      break;
   case DICT_UNKNOWN:
   default:
      break;
   }
}

int dict_data_zip( const char *inFilename, const char *outFilename,
		   const char *preFilter, const char *postFilter )
{
   char          inBuffer[IN_BUFFER_SIZE];
   char          outBuffer[OUT_BUFFER_SIZE];
   int           count;
   unsigned long inputCRC = crc32( 0L, Z_NULL, 0 );
   z_stream      zStream;
   FILE          *outStr;
   FILE          *inStr;
   int           len;
   struct stat   st;
   char          *header;
   int           headerLength;
   int           dataLength;
   int           extraLength;
   int           chunkLength;
#if HEADER_CRC
   int           headerCRC;
#endif
   unsigned long chunks;
   unsigned long chunk = 0;
   unsigned long total = 0;
   int           i;
   char          tail[8];
   char          *pt, *origFilename;

   
   /* Open files */
   if (!(inStr = fopen( inFilename, "r" )))
      err_fatal_errno( __func__,
		       "Cannot open \"%s\" for read\n", inFilename );
   if (!(outStr = fopen( outFilename, "w" )))
      err_fatal_errno( __func__,
		       "Cannot open \"%s\"for write\n", outFilename );

   origFilename = xmalloc( strlen( inFilename ) + 1 );
   if ((pt = strrchr( inFilename, '/' )))
      strcpy( origFilename, pt + 1 );
   else
      strcpy( origFilename, inFilename );

   /* Initialize compression engine */
   zStream.zalloc    = NULL;
   zStream.zfree     = NULL;
   zStream.opaque    = NULL;
   zStream.next_in   = NULL;
   zStream.avail_in  = 0;
   zStream.next_out  = NULL;
   zStream.avail_out = 0;
   if (deflateInit2( &zStream,
		     Z_BEST_COMPRESSION,
		     Z_DEFLATED,
		     -15,	/* Suppress zlib header */
		     Z_BEST_COMPRESSION,
		     Z_DEFAULT_STRATEGY ) != Z_OK)
      err_internal( __func__,
		    "Cannot initialize deflation engine: %s\n", zStream.msg );

   /* Write initial header information */
   chunkLength = (preFilter ? PREFILTER_IN_BUFFER_SIZE : IN_BUFFER_SIZE );
   fstat( fileno( inStr ), &st );
   chunks = st.st_size / chunkLength;
   if (st.st_size % chunkLength) ++chunks;
   PRINTF(DBG_VERBOSE,("%lu chunks * %u per chunk = %lu (filesize = %lu)\n",
			chunks, chunkLength, chunks * chunkLength,
			(unsigned long) st.st_size ));
   dataLength   = chunks * 2;
   extraLength  = 10 + dataLength;
   headerLength = GZ_FEXTRA_START
		  + extraLength		/* FEXTRA */
		  + strlen( origFilename ) + 1	/* FNAME  */
		  + (HEADER_CRC ? 2 : 0);	/* FHCRC  */
   PRINTF(DBG_VERBOSE,("(data = %d, extra = %d, header = %d)\n",
		       dataLength, extraLength, headerLength ));
   header = xmalloc( headerLength );
   for (i = 0; i < headerLength; i++) header[i] = 0;
   header[GZ_ID1]        = GZ_MAGIC1;
   header[GZ_ID2]        = GZ_MAGIC2;
   header[GZ_CM]         = Z_DEFLATED;
   header[GZ_FLG]        = GZ_FEXTRA | GZ_FNAME;
#if HEADER_CRC
   header[GZ_FLG]        |= GZ_FHCRC;
#endif
   header[GZ_MTIME+3]    = (st.st_mtime & 0xff000000) >> 24;
   header[GZ_MTIME+2]    = (st.st_mtime & 0x00ff0000) >> 16;
   header[GZ_MTIME+1]    = (st.st_mtime & 0x0000ff00) >>  8;
   header[GZ_MTIME+0]    = (st.st_mtime & 0x000000ff) >>  0;
   header[GZ_XFL]        = GZ_MAX;
   header[GZ_OS]         = GZ_OS_UNIX;
   header[GZ_XLEN+1]     = (extraLength & 0xff00) >> 8;
   header[GZ_XLEN+0]     = (extraLength & 0x00ff) >> 0;
   header[GZ_SI1]        = GZ_RND_S1;
   header[GZ_SI2]        = GZ_RND_S2;
   header[GZ_SUBLEN+1]   = ((extraLength - 4) & 0xff00) >> 8;
   header[GZ_SUBLEN+0]   = ((extraLength - 4) & 0x00ff) >> 0;
   header[GZ_VERSION+1]  = 0;
   header[GZ_VERSION+0]  = 1;
   header[GZ_CHUNKLEN+1] = (chunkLength & 0xff00) >> 8;
   header[GZ_CHUNKLEN+0] = (chunkLength & 0x00ff) >> 0;
   header[GZ_CHUNKCNT+1] = (chunks & 0xff00) >> 8;
   header[GZ_CHUNKCNT+0] = (chunks & 0x00ff) >> 0;
   strcpy( &header[GZ_FEXTRA_START + extraLength], origFilename );
   fwrite( header, 1, headerLength, outStr );
    
   /* Read, compress, write */
   while (!feof( inStr )) {
      if ((count = fread( inBuffer, 1, chunkLength, inStr ))) {
	 dict_data_filter( inBuffer, &count, IN_BUFFER_SIZE, preFilter );

	 inputCRC = crc32( inputCRC, inBuffer, count );
	 zStream.next_in   = inBuffer;
	 zStream.avail_in  = count;
	 zStream.next_out  = outBuffer;
	 zStream.avail_out = OUT_BUFFER_SIZE;
	 if (deflate( &zStream, Z_FULL_FLUSH ) != Z_OK)
	    err_fatal( __func__, "deflate: %s\n", zStream.msg );
	 assert( zStream.avail_in == 0 );
	 len = OUT_BUFFER_SIZE - zStream.avail_out;
	 assert( len <= 0xffff );

	 dict_data_filter( outBuffer, &len, OUT_BUFFER_SIZE, postFilter );
	 
	 assert( len <= 0xffff );
	 header[GZ_RNDDATA + chunk*2 + 1] = (len & 0xff00) >>  8;
	 header[GZ_RNDDATA + chunk*2 + 0] = (len & 0x00ff) >>  0;
	 fwrite( outBuffer, 1, len, outStr );

	 ++chunk;
	 total += count;
	 if (dbg_test( DBG_VERBOSE )) {
	    printf( "chunk %5lu: %lu of %lu total\r",
		    chunk, total, (unsigned long) st.st_size );
	    fflush( stdout );
	 }
      }
   }
   PRINTF(DBG_VERBOSE,("total: %lu chunks, %lu bytes\n", chunks, (unsigned long) st.st_size));
    
   /* Write last bit */
#if 0
   dmalloc_verify(0);
#endif
   zStream.next_in   = inBuffer;
   zStream.avail_in  = 0;
   zStream.next_out  = outBuffer;
   zStream.avail_out = OUT_BUFFER_SIZE;
   if (deflate( &zStream, Z_FINISH ) != Z_STREAM_END)
      err_fatal( __func__, "deflate: %s\n", zStream.msg );
   assert( zStream.avail_in == 0 );
   len = OUT_BUFFER_SIZE - zStream.avail_out;
   fwrite( outBuffer, 1, len, outStr );
   PRINTF(DBG_VERBOSE,("(wrote %d bytes, final, crc = %lx)\n",
		       len, inputCRC ));

   /* Write CRC and length */
#if 0
   dmalloc_verify(0);
#endif
   tail[0 + 3] = (inputCRC & 0xff000000) >> 24;
   tail[0 + 2] = (inputCRC & 0x00ff0000) >> 16;
   tail[0 + 1] = (inputCRC & 0x0000ff00) >>  8;
   tail[0 + 0] = (inputCRC & 0x000000ff) >>  0;
   tail[4 + 3] = (st.st_size & 0xff000000) >> 24;
   tail[4 + 2] = (st.st_size & 0x00ff0000) >> 16;
   tail[4 + 1] = (st.st_size & 0x0000ff00) >>  8;
   tail[4 + 0] = (st.st_size & 0x000000ff) >>  0;
   fwrite( tail, 1, 8, outStr );

   /* Write final header information */
#if 0
   dmalloc_verify(0);
#endif
   rewind( outStr );
#if HEADER_CRC
   headerCRC = crc32( 0L, Z_NULL, 0 );
   headerCRC = crc32( headerCRC, header, headerLength - 2);
   header[headerLength - 1] = (headerCRC & 0xff00) >> 8;
   header[headerLength - 2] = (headerCRC & 0x00ff) >> 0;
#endif
   fwrite( header, 1, headerLength, outStr );

   /* Close files */
#if 0
   dmalloc_verify(0);
#endif
   fclose( outStr );
   fclose( inStr );
    
   /* Shut down compression */
   if (deflateEnd( &zStream ) != Z_OK)
      err_fatal( __func__, "defalteEnd: %s\n", zStream.msg );

   xfree( origFilename );
   xfree( header );

   return 0;
}

static const char *id_string (void)
{
   static char buffer[BUFFERSIZE];
   char        *pt;

   snprintf( buffer, BUFFERSIZE, "%s", DICT_VERSION );
   pt = buffer + strlen( buffer );

   return buffer;
}

static void banner( void )
{
   fprintf( stderr, "%s %s\n", err_program_name (), id_string () );
   fprintf( stderr,
	    "Copyright 1996-2002 Rickard E. Faith (faith@dict.org)\n" );
   fprintf( stderr,
	    "Copyright 2002-2007 Aleksey Cheusov (vle@gmx.net)\n" );
   fprintf( stderr, "\n" );
}

static void license( void )
{
   static const char *license_msg[] = {
     "This program is free software; you can redistribute it and/or modify it",
     "under the terms of the GNU General Public License as published by the",
     "Free Software Foundation; either version 1, or (at your option) any",
     "later version.",
     "",
     "This program is distributed in the hope that it will be useful, but",
     "WITHOUT ANY WARRANTY; without even the implied warranty of",
     "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU",
     "General Public License for more details.",
     "",
     "You should have received a copy of the GNU General Public License along",
     "with this program; if not, write to the Free Software Foundation, Inc.,",
     "675 Mass Ave, Cambridge, MA 02139, USA.",
   0 };
   const char        **p = license_msg;
   
   banner();
   while (*p) fprintf( stderr, "   %s\n", *p++ );
}

static void help( void )
{
   static const char *help_msg[] = {
      "Usage: dictzip [options] name",
      "",
      "-d --decompress      decompress",
      "-f --force           force overwrite of output file",
      "-h --help            give this help",
      "-k --keep            do not delete original file",
      "-l --list            list compressed file contents",
      "-L --license         display software license",
      "-c --stdout          write to stdout (decompression only)",
      "-t --test            test compressed file integrity",
      "-v --verbose         verbose mode",
      "-V --version         display version number",
      "-D --debug           select debug option",
      "-s --start <offset>  starting offset for decompression (decimal)",
      "-e --size <offset>   size for decompression (decimal)",
      "-S --Start <offset>  starting offset for decompression (base64)",
      "-E --Size <offset>   size for decompression (base64)",
      "-p --pre <filter>    pre-compression filter",
      "-P --post <filter>   post-compression filter",
      0 };
   const char        **p = help_msg;

   banner();
   while (*p) fprintf( stderr, "%s\n", *p++ );
}

int main( int argc, char **argv )
{
   int           c;
   size_t        i;
   size_t        j;
   int           decompressFlag = 0;
   int           forceFlag      = 0;
   int           keepFlag       = 0;
   int           listFlag       = 0;
   int           stdoutFlag     = 0;
   int           testFlag       = 0;
   char          buffer[BUFFERSIZE];
   char          *buf;
   char          *pre           = NULL;
   char          *post          = NULL;
   unsigned long start          = 0;
   unsigned long size           = 0;
   unsigned long clSize         = 0; /* from command line */
   unsigned long clStart        = 0; /* from comment line */
   dictData      *header;
   char          *pt;
   FILE          *str;
   int           len;
   char          filename[BUFFERSIZE];
   struct option longopts[] = {
      { "stdout",       0, 0, 'c' },
      { "decompress",   0, 0, 'd' },
      { "force",        0, 0, 'f' },
      { "help",         0, 0, 'h' },
      { "keep",         0, 0, 'k' },
      { "list",         0, 0, 'l' },
      { "license",      0, 0, 'L' },
      { "test",         0, 0, 't' },
      { "verbose",      0, 0, 'v' },
      { "version",      0, 0, 'V' },
      { "debug",        1, 0, 'D' },
      { "start",        1, 0, 's' },
      { "size",         1, 0, 'e' },
      { "Start",        1, 0, 'S' },
      { "Size",         1, 0, 'E' },
      { "pre",          1, 0, 'p' },
      { "post",         1, 0, 'P' },
      { 0,              0, 0,  0  }
   };

   /* Initialize Libmaa */
   maa_init( argv[0] );
   dbg_register( DBG_VERBOSE, "verbose" );
   dbg_register( DBG_ZIP,     "zip" );
   dbg_register( DBG_UNZIP,   "unzip" );

   if ((pt = strrchr( argv[0], '/' ))) ++pt;
   else                                pt = argv[0];
   if (!strcmp(pt, "dictunzip")) ++decompressFlag;
   if (!strcmp(pt, "dictzcat")) {
      ++decompressFlag;
      ++stdoutFlag;
   }
   
#if 0
   if (signal( SIGINT, SIG_IGN ) != SIG_IGN)  signal( SIGINT, sig_handler );
   if (signal( SIGQUIT, SIG_IGN ) != SIG_IGN) signal( SIGQUIT, sig_handler );
#endif

   while ((c = getopt_long( argc, argv,
			    "cdfhklLe:E:s:S:tvVD:p:P:",
			    longopts, NULL )) != EOF)
      switch (c) {
      case 'd': ++decompressFlag;                                      break;
      case 'f': ++forceFlag;                                           break;
      case 'k': ++keepFlag;                                            break;
      case 'l': ++listFlag;                                            break;
      case 'L': license(); exit( 1 );                                  break;
      case 'c': ++stdoutFlag;                                          break;
      case 't': ++testFlag;                                            break;
      case 'v': dbg_set( "verbose" );                                  break;
      case 'V': banner(); exit( 1 );                                   break;
      case 'D': dbg_set( optarg );                                     break;
      case 's': ++decompressFlag; clStart = strtoul( optarg, NULL, 10 ); break;
      case 'e': ++decompressFlag; clSize  = strtoul( optarg, NULL, 10 ); break;
      case 'S': ++decompressFlag; clStart = b64_decode( optarg );        break;
      case 'E': ++decompressFlag; clSize  = b64_decode( optarg );        break;
      case 'p': pre = optarg;                                          break;
      case 'P': post = optarg;                                         break;
      default:  
      case 'h': help(); exit( 1 );                                     break;
      }

   if (testFlag) ++listFlag;

   for (i = optind; i < (size_t) argc; i++) {
      size  = clSize  ? clSize  : 0;
      start = clStart ? clStart : 0;
      if (listFlag) {
	 header = dict_data_open( argv[i], 1 );
	 dict_data_print_header( stdout, header );
	 dict_data_close( header );
      } else if (decompressFlag) {
	 if (stdoutFlag) {
	    header = dict_data_open( argv[i], 0 );
	    if (!size) size = header->length;
	    if (!start) {
	       len = header->chunkLength;
	       for (j = 0; j < size; j += len) {
		  if (j + len >= size) len = size - j;
		  buf = dict_data_read_ ( header, j, len, pre, post );
		  fwrite( buf, len, 1, stdout );
		  fflush( stdout );
		  xfree( buf );
	       }
	    } else {
	       buf = dict_data_read_ ( header, start, size, pre, post );
	       fwrite( buf, size, 1, stdout );
	       fflush( stdout );
	       xfree( buf );
	    }
	    dict_data_close( header );
	 } else {
	    strlcpy( filename, argv[i], BUFFERSIZE );

	    if ((pt = strrchr( filename, '.' ))) *pt = '\0';
	    else
	       err_fatal( __func__, "Cannot truncate filename\n" );
	    if (!forceFlag && (str = fopen( filename, "r" )))
	       err_fatal( __func__, "%s already exists\n", filename );
	    if (!(str = fopen( filename, "w" )))
	       err_fatal_errno( __func__,
				"Cannot open %s for write\n", filename );
	    header = dict_data_open( argv[i], 0 );
	    if (!size) size = header->length;
	    len = header->chunkLength;
	    for (j = 0; j < size; j += len) {
	       if (j + len >= size) len = size - j;
	       buf = dict_data_read_ ( header, j, len, pre, post );
	       fwrite( buf, len, 1, str );
	       fflush( str );
	       xfree( buf );
	    }	    
	    dict_data_close( header );
	    if (!keepFlag && unlink( argv[i] ))
	       err_fatal_errno( __func__, "Cannot unlink %s\n", argv[i] );
	 }
      } else {
	 snprintf( buffer,BUFFERSIZE-1, "%s.dz", argv[i] );
	 if (!dict_data_zip( argv[i], buffer, pre, post )) {
	    if (!keepFlag && unlink( argv[i] ))
		err_fatal_errno( __func__, "Cannot unlink %s\n", argv[i] );
	 } else {
	    err_fatal( __func__, "Compression failed\n" );
	 }
      }
   }

   return 0;
}
